### ISSUE TITLE
useField isChecked value

### ISSUE THREAD

**Issue by souldreamer** [Sawtaytoes/OneForm#9]
Trying to figure out what type the `isChecked` field in the return value of `useField` actually is. However, it's computed as:

```typescript
		isChecked: isCheckboxElement ? Boolean(value) : value === inputValue ? value : false,
```

Problem is that it doesn't guarantee booleans (which is a bit counter-intuitive), but only in the case the value hasn't changed from its initial value? Can you please explain the reasoning here and if it actually serves any purpose to not have it only be boolean?

**Comment by Sawtaytoes** [comment]
`isChecked` should be only either `true` or `false` by convention, but I might've named it wrong.

I shouldn't have a nested ternary there. Meant to clean it up and forgot.

Checkboxes don't guarantee booleans in HTML. There's an `'indeterminate'` state as well:

```jsx
<input
  checked={/* either `'true'` or `null` */}
  indeterminate={/* either `'true'` or `null` */}
  value={/* either `'true'`, `null`, or some string */}`
/>
```

If we have no `value` defined on the checkbox, it's a boolean.

When interacting with HTML, we need to convert it to a string `'true'` to denote it has a value. Another common practice is setting the prop to its own name: `checked="checked"`.

If `value` is defined, we need to store it (`inputValue`) instead of the boolean state.

---

From what I'm seeing, `isChecked` will always be a boolean. The first check forces it to a boolean value. The second check, `value === inputValue`, can never be true for a checkbox as it's checking `boolean === string`.

If the input isn't a checkbox type, then `value` is null`. If `null === ''`, we get `false`.

**Comment by souldreamer** [comment]
> If the input isn't a checkbox type, then you're right, we might end up setting it to a non-boolean value.

Shouldn't we try to keep it a boolean no matter what? It's just easier to reason about for users, and this way they get no surprise values for it.

**Comment by Sawtaytoes** [comment]
Correct. I'll have to look at `Field` to make sure, but this might be a place where the entire output value could be wrapped in `Boolean`. We might need to look for `0` first, but I don't think so.

If you're caring about `isChecked`, your `value` will be a boolean or string. It's DOM-specific meaning, we can assume no numbers are passed. But... Someone _could_ pass a number as a default value and that would break the whole thing. We should probably adjust for `0`.

**Comment by souldreamer** [comment]
Actually, with component libraries and more complex components, `string`s and `number`s are definitely not the only possible value types people will try to control with `OneForm`.

For example, the component library I'm using for my current project is `primereact`, and quite a few select-type inputs ([listbox](https://primefaces.org/primereact/showcase/#/listbox), multi-select, even select-button) actually expect `object` values. It's the reason I left the value typing as `any` in the typings file, and didn't force it to `string` or `string | number`. Have to remember that forms in react rarely deal with actual html inputs directly nowadays, it's usually a component library of some sort.

**Comment by Sawtaytoes** [comment]
Fixed in: bb1976da11a44ad7e57c0b0af7c0cd36f5af8ae1.

We're gonna have to figure out a better way of doing this nested ternary, but the solution today works.
